### PR TITLE
[BTAT-11289] Read breathingSpace block from 1812

### PR DESCRIPTION
**Pull request by daparkes**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -227,6 +227,11 @@ Response Body:
 |LPP2Percentage|`Double`|**false**|    
 |penaltyChargeReference|`String`|**false**|
 
+##### BreathingSpace object
```
Comment (lewishay) [inline README.md]: Breathing space should be a field at the same level of LPPDetails, not a separate object

File: app/models/API1812/BreathingSpace.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.time.LocalDate
+
+import play.api.libs.json.{Format, Json}
+
+case class BreathingSpace(BSStartDate: LocalDate,
+                          BSEndDate: LocalDate)
+
+object BreathingSpace {
+  implicit val formats: Format[BreathingSpace] = Json.format[BreathingSpace]
```
Comment (lewishay) [inline app/models/API1812/BreathingSpace.scala]: Can you change this to `Reads` as we don't need writing functionality for these dates

File: app/services/API1812/PenaltyDetailsService.scala [hunk 3]
```diff
@@ -31,8 +29,5 @@ class PenaltyDetailsService @Inject()(connector: PenaltyDetailsConnector) extend
 
   def getPenaltyDetails(regime: TaxRegime, queryParameters: PenaltyDetailsQueryParameters)
                        (implicit hc: HeaderCarrier, ec: ExecutionContext): Future[PenaltyDetailsResponse] =
-    connector.getPenaltyDetails(regime, queryParameters).map {
-      case Right(details) if details.LPPDetails.isEmpty => Left(Error(Status.NOT_FOUND, "No LPP data was found"))
-      case either => either
-    }
+    connector.getPenaltyDetails(regime, queryParameters)
```
Comment (lewishay) [inline app/services/API1812/PenaltyDetailsService.scala:32]: Following this change can you add in a new IT test in the 1812 component spec to confirm that when the API gives us a 200 response with empty JSON, we will write JSON to the frontend as expected (`LPPDetails` as an empty array and breathing space as false)

File: app/models/API1812/PenaltyDetails.scala [hunk 4]
```diff
@@ -16,17 +16,34 @@
...
-import play.api.libs.json.{Reads, Writes, __}
+import play.api.libs.functional.syntax.toFunctionalBuilderOps
+import play.api.libs.json.{Json, Reads, Writes, __}
+import services.DateService
 
-case class PenaltyDetails(LPPDetails: Option[Seq[LatePaymentPenalty]])
+case class PenaltyDetails(LPPDetails: Option[Seq[LatePaymentPenalty]],
+                          breathingSpace: Option[BreathingSpace]) {
+
+  def hasBreathingSpace(implicit appConfig: AppConfig): Boolean = breathingSpace match {
```
Comment (lewishay) [inline app/models/API1812/PenaltyDetails.scala]: Can you please add unit tests for this function. 
Seems like this would incorrectly return false if the user was on the first day of BS? That scenario and other boundary cases would be good candidates for the unit tests.

File: README.md [hunk 5]
```diff
@@ -249,29 +254,16 @@ Response Body:
             "penaltyChargeReference": "BCDEFGHIJKLMNOPQ"    
         }    
     ]    
-}    
+},
+{
```
Comment (lewishay) [inline README.md]: Same sort of comment as above, this JSON is currently invalid as you have two top-level JSON objects

**Review by lewishay — CHANGES_REQUESTED** [review]

File: app/models/API1812/PenaltyDetails.scala [hunk 6]
```diff
@@ -16,17 +16,34 @@
...
 import models.API1812.latePaymentPenalty.LatePaymentPenalty
-import play.api.libs.functional.syntax.unlift
-import play.api.libs.json.{Reads, Writes, __}
+import play.api.libs.functional.syntax.toFunctionalBuilderOps
+import play.api.libs.json.{Json, Reads, Writes, __}
+import services.DateService
 
-case class PenaltyDetails(LPPDetails: Option[Seq[LatePaymentPenalty]])
+case class PenaltyDetails(LPPDetails: Option[Seq[LatePaymentPenalty]],
+                          breathingSpace: Option[BreathingSpace]) {
```
Comment (lewishay) [inline app/models/API1812/PenaltyDetails.scala]: Breathing space is actually an array, so there could be multiple sets of BS. Unfortunately that means a bit of extra logic as you'll have to check if the user is in BS for any of the periods.

File: it/controllers/PenaltyDetailsComponentSpec.scala [hunk 7]
```diff
@@ -52,6 +52,28 @@ class PenaltyDetailsComponentSpec extends ComponentSpecBase {
       }
     }
 
+    "the API returns a success response with breathing space but no penalty data" should {
```
Comment (lewishay) [inline it/controllers/PenaltyDetailsComponentSpec.scala:55]: The scenario I was thinking of was when we get a 200 response and completely empty JSON, as EIS have told us this is to be expected and will be fairly common, i.e. for users with no penalties or breathing space info. We can leave this test in now you've written it, but would still think it's beneficial for us to have a test for the empty JSON scenario.

File: it/testData/PenaltyDetailsTestData.scala [hunk 8]
```diff
@@ -38,13 +39,29 @@ object PenaltyDetailsTestData {
...
+  val penaltyDetailsAPIJsonBSOnly: JsObject = Json.obj(
+    "breathingSpace" -> Json.arr(breathingSpaceJson)
+  )
+
+  val penaltyDetailsWrittenJson: JsObject = Json.obj(
+    "LPPDetails" -> Json.arr(LPPJson), "breathingSpace" -> false
+  )
+
+  val penaltyDetailsWrittenJsonBSOnly: JsObject = Json.obj(
+    "LPPDetails" -> JsNull, "breathingSpace" -> false
```
Comment (lewishay) [inline it/testData/PenaltyDetailsTestData.scala]: We want this to be an empty array if there are no penalty details. A `null` value will fail to parse on the frontend as it expects `LPPDetails` to be of type array, not null. Can you tweak the json writes to accomodate this. Looks like it may be as simple as using a `.getOrElse(Seq())` on the `LPPDetails`?

File: test/models/API1812/PenaltyDetailsSpec.scala [hunk 9]
```diff
@@ -29,25 +29,84 @@ class PenaltyDetailsSpec extends SpecBase {
...
+
+      "user is well within breathing space" in {
+        PenaltyDetails(Some(Seq(LPPModelMin)), Some(Seq(inBS))).hasBreathingSpace shouldBe true
+      }
+
+      "user is on the first day of breathing space" in {
+        PenaltyDetails(Some(Seq(LPPModelMin)), Some(Seq(firstDayBS))).hasBreathingSpace shouldBe true
+      }
+
+      "user is on the last day of breathing space" in {
```
Comment (lewishay) [inline test/models/API1812/PenaltyDetailsSpec.scala:77]: This is an interesting case to me. If their BS ends on 2018-01-05, and today is 2018-01-05, I would have assumed this would be `false`. I suppose it could be either though. I've asked the question to Martin/Cheryl/Michael and will let you know the outcome.

**Review by lewishay — CHANGES_REQUESTED** [review]

**Review by lewishay — APPROVED** [review]
